nixos/modules/system/activation/switch-to-configuration.pl

```diff
@@ -889,10 +889,6 @@ sub filter_units {
 
 close($list_active_users) || die("Unable to close the file handle to loginctl");
 
-# Set the new tmpfiles
-print STDERR "setting up tmpfiles\n";
-system("$new_systemd/bin/systemd-tmpfiles", "--create", "--remove", "--exclude-prefix=/dev") == 0 or $res = 3;
-
```

Comment:
So this implicitly always ran before any units told to "refresh" themselves, or start for the first time.
Many services use `systemd.tmpfiles.rules` to create a directory that the service needs, so this must run before any services are started, (or - I suppose - restarted or reloaded).
How can a service ensure that `systemd-tmpfiles` runs before it starts?
I don't think we have any mechanism for applying an ordering to the "switch" operations on units.
Until we have something for that, perhaps the new unit should always be (re)started before all other units?

Reply:
(off-topic, because we can't change all services overnight:)
I do think services should just add an `ExecStart` that runs as root and takes care of directory setup, if they can't use the systemd `StateDirectory` etc. It's more self-contained and robust that way.

Unfortunately `ExecStart` is unpleasant to work with. I think we should either
- have https://github.com/NixOS/nixpkgs/pull/154154
- or remove `systemd.services.<name>.script`.

Reply:
Yes this already catastrophically exploded in our faces: https://github.com/NixOS/nixpkgs/pull/269983
I thought that because tmpfiles is ordered before sysinit.target this would all work out correctly. 
However this is not the case.

Overall, stc orders units correctly but not necessarily across target boundaries. And apparently sysinit.target is a boundary we cannot overcome right now.

 I think I'll just add an (re-)activationScript that runs systemd-sysusers before systemd-tmpfiles

Reply:
I significantly reworked the previous PR: https://github.com/NixOS/nixpkgs/pull/271067